pkg/streamingpromql/operators/function_over_range_vector.go

```diff
@@ -107,18 +118,117 @@ func (m *FunctionOverRangeVector) NextSeries(ctx context.Context) (types.Instant
 			}
 		}
 
-		accumulate(head)
-		accumulate(tail)
+		accumulate(fHead)
+		accumulate(fTail)
+
+		val := m.calculateFloatRate(step.RangeStart, step.RangeEnd, firstPoint, lastPoint, delta, fCount)
+		if data.Floats == nil {
+			// Only get fPoint slice once we are sure we have float points.
+			// This potentially over-allocates as some points in the steps may be histograms,
+			// but this is expected to be rare.
+			data.Floats, err = m.Pool.GetFPointSlice(m.numSteps)
+			if err != nil {
+				return err
+			}
+		}
+		data.Floats = append(data.Floats, promql.FPoint{T: step.StepT, F: val})
+	}
 
-		val := m.calculateRate(step.RangeStart, step.RangeEnd, firstPoint, lastPoint, delta, count)
+	if hCount >= 2 {
+		firstPoint := m.histogramBuffer.First()
+		lastPoint, _ := m.histogramBuffer.LastAtOrBefore(step.RangeEnd) // We already know there is a point at or before this time, no need to check.
 
-		data.Floats = append(data.Floats, promql.FPoint{T: step.StepT, F: val})
+		currentSchema := firstPoint.H.Schema
+		if lastPoint.H.Schema < currentSchema {
+			currentSchema = lastPoint.H.Schema
+		}
+
+		delta := lastPoint.H.CopyToSchema(currentSchema)
+		_, err = delta.Sub(firstPoint.H)
+		if err != nil {
+			return err
+		}
+		previousValue := firstPoint.H
+
+		accumulate := func(points []promql.HPoint) error {
+			for _, p := range points {
+				if p.T > step.RangeEnd { // The buffer is already guaranteed to only contain points >= rangeStart.
+					return nil
+				}
+
```

Comment:
I think this was a leftover from before `FPointRingBuffer.UnsafePoints` took a `maxT` parameter - this `if` should be safe to remove here and in `accumulate` for float values above too.